### PR TITLE
refactor(medium-pass-1): add forEachConfig and medium pass provider to VisualizationConfigurationFactory

### PR DESCRIPTION
**Pull request by waabid**
#### Details

Heavily suggest reviewing by commit due to signature changes causing many files to be changed.

This PR can roughly be broken into 3 things:

1. Use constructor to pass in assessment and medium pass provider: this requires the most amount of file changes as each usage/creation of the factory needs to be changed.
2. Add the ability to loop through each configuration from the factory instead of components being responsible for this. The benefit here is that each component doing this no longer needs to be aware of the difference between assessment and not but also the structure of the assessment data.
3. Consumption of this forEachConfig method in the appropriate locations.

##### Motivation

feature work.

##### Context

1. `getIdentifier` is refactored out of AssessmentBuilder. This allows us to use test mode as a part of generating the identifier allowing for it to be unique across medium pass and assessment; removing it is the 3rd commit but adding it to config factory is in the first commit.
2. The mediumPass provider is currently optional so as not to cause issues/errors as the work on that is in the next PR. This will be removed in that PR as well.
3. I did verify that the async behavior in `VisualizationStateChangeHandler` was kept the same.
4. There is obviously a lot of potential refactoring work around the configuration factory (many tests don't mock it, we could build configs only once, we could break the building from the using/getting, etc.) but this PR is large enough as it is and I tried to maintain what we were doing in the main branch as best as I could.

#### Pull request checklist

- [ ] Addresses an existing issue: #0000
- [x] Ran `yarn null:autoadd`
- [x] Ran `yarn fastpass`
- [x] Added/updated relevant unit test(s) (and ran `yarn test`)
- [x] Verified code coverage for the changes made. Check coverage report at: `<rootDir>/test-results/unit/coverage`
- [x] PR title *AND* final merge commit title both start with a semantic tag (`fix:`, `chore:`, `feat(feature-name):`, `refactor:`). See `CONTRIBUTING.md`.
- [ ] (UI changes only) Added screenshots/GIFs to description above
- [ ] (UI changes only) Verified usability with NVDA/JAWS

### CODE REVIEW

File: src/injected/visualization-state-change-handler.ts [hunk 1]
```diff
@@ -1,35 +1,30 @@
...
-                        ),
-                    );
-                } else {
-                    await this.visualizationUpdater(visualizationType, null, storeData);
-                }
-            }),
+
+        const updateCalls = [];
+        this.visualizationConfigurationFactory.forEachConfig(
+            (testConfig, type, requirementConfig) => {
```
Comment (madalynrose) [inline src/injected/visualization-state-change-handler.ts:21]:
```suggestion
            (testConfig: VisualizationConfiguration, type: VisualizationType, requirementConfig: Requirement) => {
```

Explicit types increase readability.

File: src/injected/analyzer-controller.ts [hunk 2]
```diff
@@ -82,27 +78,22 @@ export class AnalyzerController {
...
-                        stepConfig.key,
-                    );
-                });
-            } else {
-                const key = config.getIdentifier();
-                this.analyzers[key] = config.getAnalyzer(this.analyzerProvider);
-            }
-        });
+        this.visualizationConfigurationFactory.forEachConfig(
+            (testConfig, type, requirementConfig) => {
```
Comment (madalynrose) [inline src/injected/analyzer-controller.ts:82]:
```suggestion
            (testConfig: VisualizationConfiguration, type: VisualizationType, requirementConfig: Requirement) => { => {
```

Explicit types increase readability.

File: src/injected/analyzer-state-update-handler.ts [hunk 3]
```diff
@@ -41,19 +35,27 @@ export class AnalyzerStateUpdateHandler {
...
-                    }
-                });
-            }
-        });
+        if (prevState == null) {
+            return;
+        }
+
+        this.visualizationConfigurationFactory.forEachConfig(
+            (configuration, type, requirementConfig) => {
```
Comment (madalynrose) [inline src/injected/analyzer-state-update-handler.ts:43]:
```suggestion
           (configuration: VisualizationConfiguration, type: VisualizationType, requirementConfig: Requirement) => {
```

Explicit types increase readability.

File: src/injected/target-page-visualization-updater.ts [hunk 4]
```diff
@@ -54,7 +54,7 @@ export class TargetPageVisualizationUpdater {
         const configuration =
             this.visualizationConfigurationFactory.getConfiguration(visualizationType);
         const configId = configuration.getIdentifier(stepKey);
-
+        console.log('type is ', visualizationType);
```
Comment (madalynrose) [inline src/injected/target-page-visualization-updater.ts:57]:
```suggestion
```

Remove console.log

File: src/common/configs/web-visualization-configuration-factory.ts [hunk 5]
```diff
@@ -65,17 +61,67 @@ export class WebVisualizationConfigurationFactory implements VisualizationConfig
...
-            const configuration = this.configurationByType[visualizationType];
+    private buildAssessmentConfiguration(
+        assessment: Assessment,
+        testMode: TestMode,
+    ): VisualizationConfiguration {
+        const config = assessment.getVisualizationConfiguration();
 
-            if (configuration && configuration.chromeCommand != null) {
-                map[configuration.chromeCommand] = visualizationType;
+        const getIdentifier = step => {
```
Comment (madalynrose) [inline src/common/configs/web-visualization-configuration-factory.ts:92]:
```suggestion
        const getIdentifier = testStep: string => {
```

Explicit types increase readability, changing variable name to match declaration in `VisualizationConfiguration`.

**Review by madalynrose — COMMENTED** [review]
A few small comments (I tried to make suggestions for nits so you can just press a button instead of having to go through and make small changes) but overall looks good!

File: src/common/configs/web-visualization-configuration-factory.ts [hunk 6]
```diff
@@ -65,17 +61,67 @@ export class WebVisualizationConfigurationFactory implements VisualizationConfig
         return configuration;
     }
 
-    public getChromeCommandToVisualizationTypeMap(): DictionaryStringTo<VisualizationType> {
-        const map: DictionaryStringTo<VisualizationType> = {};
+    public forEachConfig(callback: ForEachConfigCallback): void {
+        Object.keys(this.configurationByType).map(type => {
```
[the remark this answers is not in the document]
Comment (waabid) [reply src/common/configs/web-visualization-configuration-factory.ts:65]: Good catch; this is a remnant from a previous change!

**Review by madalynrose — APPROVED** [review]
LGTM!

**Pull request merged**